kube-runtime/src/controller/mod.rs

```diff
@@ -206,6 +206,51 @@ impl Display for ReconcileReason {
 
 const APPLIER_REQUEUE_BUF_SIZE: usize = 100;
 
+pub trait Reconciler<T, ReconcilerFut, K, Ctx>
+where
+    K: Resource,
+{
+    fn call(&mut self, obj: Arc<K>, ctx: Arc<Ctx>, request: ReconcileRequest<K>) -> ReconcilerFut;
+}
+
+impl<ReconcilerFut, K, Ctx, F> Reconciler<(Arc<K>, Arc<Ctx>), ReconcilerFut, K, Ctx> for F
+where
+    F: FnMut(Arc<K>, Arc<Ctx>) -> ReconcilerFut,
+    K: Resource,
+{
+    fn call(&mut self, obj: Arc<K>, ctx: Arc<Ctx>, _: ReconcileRequest<K>) -> ReconcilerFut {
+        self(obj, ctx)
+    }
+}
+
+impl<ReconcilerFut, K, Ctx, F>
+    Reconciler<(Arc<K>, Arc<Ctx>, Option<Box<ObjectRef<DynamicObject>>>), ReconcilerFut, K, Ctx> for F
+where
+    F: FnMut(Arc<K>, Arc<Ctx>, Option<Box<ObjectRef<DynamicObject>>>) -> ReconcilerFut,
+    K: Resource,
+{
+    fn call(&mut self, obj: Arc<K>, ctx: Arc<Ctx>, request: ReconcileRequest<K>) -> ReconcilerFut {
```

Comment:
Disregarding the question of whether we want this at all (which is being discussed in #1091), I think I'd rather only keep the variant that takes `ReconcileRequest<K>`. It's not very clear from the type signature alone what the `ObjectRef` refers to.